### PR TITLE
Space - Jeta and Stephanie

### PR DESCRIPTION
**Pull request by stephaniejmars**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | Having a set of instructions to find specific examples of things to look for and find out was very helpful to get familiar with the files. 
What _inheritance_ relations exist between classes? | Passenger, Driver, and Trip all inherit from their parent class CSV Record
What _composition_ relations exist between classes? | Passenger to Trip, one to many. Driver to Trip is one to many. Driver to Passenger is many to many. 
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We were not sure if the trip cost was 1.65 or less, if the company kept the fee, or if the driver got the pay. We decided to give it to the company because typically that is was Corporate America would do, and we work for the company not the driver, but also, damn the man. 
Give an example of a _template method_ that you implemented for this assignment | self.build_path was a prebuilt template. It does not get called within the CSVRecord file, but other methods will call it.
Give an example of a _nominal test_ that you wrote for this assignment | expect trip must be kind of trip
Give an example of an _edge case test_ that you wrote for this assignment | if all drivers are unavailable 
What is a concept that you gained more clarity on as you worked on this assignment | Classes, Inherited, calling files from other files, writing test.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      if vin.class == Integer
+        vin.to_s
+      end
+
+      valid_statuses = [:AVAILABLE, :UNAVAILABLE]
+      raise ArgumentError if !valid_statuses.include?(@status)
+
+      raise ArgumentError if id <= 0 || id.nil?
+
+      raise ArgumentError if vin.length != 17
```
Comment (kaidamasaki) [inline lib/driver.rb:32]: Remember, error messages are helpful for your users and for debugging.

```suggestion
      raise ArgumentError, "Bad status: #{status}" if !valid_statuses.include?(@status)

      raise ArgumentError, "Bad id: #{id}" if id <= 0 || id.nil?

      raise ArgumentError, "Bad vin: #{vin}" if vin.length != 17
```

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    def add_trip(trip)
+      trips << trip
+    end
+    
+    def average_rating
+      total_ratings = 0
+      @trips.each do |trip|
+        total_ratings += trip.rating
+      end
```
Comment (kaidamasaki) [inline lib/driver.rb:43]: You can simplify this using `sum`:

```suggestion
      total_ratings = @trips.sum { |trip| trip.rating }
```

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      end 
+      @trips.each do |trip|
+        if trip.cost <= 1.65
+          total_revenue += 0
+        else
+          trip_pay = trip.cost - 1.65
+          trip_pay = trip_pay * 0.80
+          total_revenue += trip_pay
+        end
+      end
```
Comment (kaidamasaki) [inline lib/driver.rb:64]: Using `each` makes sense here.  The logic is kind of complicated.

In case you're curious about the `Enumerable` version:
```suggestion
      total_revenue = @trips.select { |trip| trip.cost > 1.65 }.sum { |trip| (trip.cost - 1.65) * 0.8 }
```

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      return total_revenue
+    end
+
+    def update_status 
+      @status = :UNAVAILABLE
+    end
```
Comment (kaidamasaki) [inline lib/driver.rb:70]: `update_status` is misleading since this method only ever sets `@status` to one value. 

I would probably name this something like `mark_unavailable` or just omit this method entirely since you already have an `attr_accessor` on `@status`.

File: lib/trip.rb [hunk 5]
```diff
@@ -1,41 +1,71 @@
...
+          end_time: nil,
           cost: nil,
-          rating:
+          rating: nil
         )
       super(id)
 
+      
+      if driver
+        @driver = driver
```
Comment (kaidamasaki) [inline lib/trip.rb:25]: Since you're only really using `driver_id` I wouldn't bother storing `@driver` as well since it could be confusing.

File: lib/trip.rb [hunk 6]
```diff
@@ -46,22 +76,26 @@ def inspect
...
 
     def self.from_csv(record)
       return self.new(
                id: record[:id],
+               driver_id: record[:driver_id],
                passenger_id: record[:passenger_id],
-               start_time: record[:start_time],
-               end_time: record[:end_time],
+               start_time: record[:start_time], # nn to turn into Time class
+               end_time: record[:end_time], # nn to turn into Time class  before passing into initialize
```
Comment (kaidamasaki) [inline lib/trip.rb:94]: You should use `Time.parse` here:

```suggestion
               start_time: Time.parse(record[:start_time]),
               end_time: Time.parse(record[:end_time]),
```

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -27,14 +34,38 @@ def inspect
               #{passengers.count} passengers>"
     end
 
+    def request_trip(passenger_id)
+      passenger = find_passenger(passenger_id)
+      driver = @drivers.select {|driver| driver.status == :AVAILABLE}.first
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:39]: You could make this clearer using `find`:

```suggestion
      driver = @drivers.find {|driver| driver.status == :AVAILABLE}
```

File: lib/trip_dispatcher.rb [hunk 8]
```diff
@@ -27,14 +34,38 @@ def inspect
...
+      raise ArgumentError, "There are no available drivers" if driver == nil
+ 
+      new_trip = Trip.new(
+        id: @trips.length + 1,
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:43]: Nice way of assigning `id`s for trips.  😄

File: lib/trip_dispatcher.rb [hunk 9]
```diff
@@ -27,14 +34,38 @@ def inspect
...
+        passenger: passenger,
+        start_time: Time.now, 
+        end_time: nil,
+        cost: nil,
+        rating: nil
+    )
+
+      new_trip.connect(passenger,driver)
+
+      driver.update_status
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:54]: You can just use the accessor here (see above):

```suggestion
      driver.status = :UNAVAILABLE
```

File: lib/trip.rb [hunk 10]
```diff
@@ -1,41 +1,71 @@
...
         @passenger_id = passenger_id
-
       else
         raise ArgumentError, 'Passenger or passenger_id is required'
       end
 
+
+      if start_time.class == String 
+        start_time = Time.parse(start_time)
+      end
```
Comment (kaidamasaki) [inline lib/trip.rb:46]: If you parse `Time`s in `from_csv` you can omit this logic.

```suggestion
```

File: lib/trip.rb [hunk 11]
```diff
@@ -1,41 +1,71 @@
...
       @start_time = start_time
+
+      if end_time.class == String 
+        end_time = Time.parse(end_time)
+      end
```
Comment (kaidamasaki) [inline lib/trip.rb:51]: If you parse Times in from_csv you can omit this logic.

```suggestion
```

File: lib/trip.rb [hunk 12]
```diff
@@ -1,41 +1,71 @@
...
       end
+
+      if @end_time && (end_time < start_time)
+        raise ArgumentError.new("End time #{end_time} is before  start time #{start_time}")
+      end
+    end 
+
+    def trip_duration_in_seconds
+      trip_duration = @end_time - @start_time
+      return trip_duration
```
Comment (kaidamasaki) [inline lib/trip.rb:68]: You don't need to assign your results to variables before returning them:

```suggestion
      return @end_time - @start_time
```

File: test/driver_test.rb [hunk 13]
```diff
@@ -131,6 +130,73 @@
...
+        end_time: Time.new(2016, 8, 8),
+        cost: 5,
+        rating: 5
+      )
+      @driver.add_trip(trip2)
+      
+    end
+
+    it "correctly calculates total_revenue" do
+      expect(@driver.total_revenue).must_be_close_to ((5-1.65) + (10-1.65)) * 0.80, 0.01
```
Comment (kaidamasaki) [inline test/driver_test.rb:164]: Since you're asserting on things set up in your before it might be nice to have those costs stored in instance variables like `@trip_cost1` and `@trip_cost2` that way if you change the cost your test will update automatically.

File: test/driver_test.rb [hunk 14]
```diff
@@ -131,6 +130,73 @@
...
+    end
+
+    it "returns zero if no driven trips" do
+      driver = RideShare::Driver.new(
+        id: 54,
+        name: "Rogers Bartell IV",
+        vin: "1C9EVBRM0YBC564DZ"
+      )
+      expect(driver.total_revenue).must_equal 0
+      expect(driver.total_revenue).must_be_kind_of Numeric
```
Comment (kaidamasaki) [inline test/driver_test.rb:199]: You don't need to test that `total_revenue` is both `Numeric` and equal to `0` since `0` is itself `Numeric`.

```suggestion
      expect(driver.total_revenue).must_equal 0
```

File: test/passenger_test.rb [hunk 15]
```diff
@@ -67,8 +74,96 @@
...
+        @trip2 = RideShare::Trip.new(
+        id: 9,
+        driver: @driver,
+        passenger: @passenger,
+        start_time: Time.new(2016, 8, 11),
+        end_time: Time.new(2016, 8, 12),
+        cost: 11,
+        rating: 5
+        )
+        @passenger.add_trip(@trip2)
```
Comment (kaidamasaki) [inline test/passenger_test.rb:115]: Incorrect indentation:

```suggestion
      @driver = RideShare::Driver.new(
        id: 7,
        name: "Merl Glover Driver",
        vin: "16026202330372367",
        status: :AVAILABLE,
        trips: []
      )
      @passenger = RideShare::Passenger.new(
        id: 9,
        name: "Merl Glover III",
        phone_number: "1-602-620-2330 x3723",
        trips: []
      )
        
      @trip1 = RideShare::Trip.new(
        id: 8,
        driver: @driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 8),
        end_time: Time.new(2016, 8, 9),
        cost: 10,
        rating: 5
      )
      @passenger.add_trip(@trip1)
      
      @trip2 = RideShare::Trip.new(
        id: 9,
        driver: @driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 11),
        end_time: Time.new(2016, 8, 12),
        cost: 11,
        rating: 5
      )
      @passenger.add_trip(@trip2)
```

File: test/passenger_test.rb [hunk 16]
```diff
@@ -67,8 +74,96 @@
...
+        @trip2 = RideShare::Trip.new(
+        id: 9,
+        driver: @driver,
+        passenger: @passenger,
+        start_time: Time.new(2016, 8, 11, 20, 00, 00),
+        end_time: Time.new(2016, 8, 11, 20, 8, 00),
+        cost: 11,
+        rating: 5
+        )
+        @passenger.add_trip(@trip2)
```
Comment (kaidamasaki) [inline test/passenger_test.rb:161]: Incorrect indentation (see above).

Also if you'd like to share a `before` block between two `describe` blocks you can move it up to their parent `describe` block.  (Or create a new parent `describe` block.)

File: test/trip_dispatch_test.rb [hunk 17]
```diff
@@ -23,15 +23,12 @@ def build_test_dispatcher
 
       expect(dispatcher.trips).must_be_kind_of Array
       expect(dispatcher.passengers).must_be_kind_of Array
-      # expect(dispatcher.drivers).must_be_kind_of Array
```
Comment (kaidamasaki) [inline test/trip_dispatch_test.rb:26]: I'm not clear on why you deleted this.  The provided tests are part of the requirements.

File: test/trip_dispatch_test.rb [hunk 18]
```diff
@@ -122,4 +119,49 @@ def build_test_dispatcher
...
+    end
+
+    it "creates a new trip" do
+      expect(@trip).must_be_kind_of RideShare::Trip
+      expect(@trip.id).must_be_instance_of Integer
+      expect(@trip.driver_id).must_equal 2
+      expect(@trip.start_time).must_be_instance_of Time
+      expect(@trip.end_time).must_be_nil
+      expect(@trip.cost).must_be_nil
+      expect(@trip.rating).must_be_nil
```
Comment (kaidamasaki) [inline test/trip_dispatch_test.rb:137]: Good use of `must_be_nil`!

File: test/trip_dispatch_test.rb [hunk 19]
```diff
@@ -122,4 +119,49 @@ def build_test_dispatcher
...
+
+    it "returns a new trip" do
+      expect(@trip).must_equal @trip
+    end
+
+    it "raises ArgumentError if all drivers are unavailable" do
+      @dispatcher.drivers.each do |driver|
+           driver.status = :UNAVAILABLE
+      end
+      expect{@dispatcher.request_trip(3)}.must_raise ArgumentError
```
Comment (kaidamasaki) [inline test/trip_dispatch_test.rb:164]: Adding some spacing can make this read a little clearer.

```suggestion
      expect { @dispatcher.request_trip(3) }.must_raise ArgumentError
```

**Comment by kaidamasaki** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | Parsed `Time`s in `Trip.initialize`.
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | Missing.
Practices inheritance. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | Your messages were fine but you should commit more often.

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | 
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️ (You tested the nominal case in `before`, but testing a nominal case explicitly is best practices.)

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements |
| Yellow (Approaches Standards) | 6+ in Code Review && 2+ in Functional Requirements | ✔️
| Red (Not at Standard) | 0-5 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |


### Additional Feedback


Generally this assignment was pretty reasonable.  Your issues here were mostly around attention to detail.  You got dinged for a lot of little things but overall you made reasonable decisions and your code was well organized.



This was awesome:

>We decided to give it to the company because typically that is was Corporate America would do, and we work for the company not the driver, but also, damn the man.




## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | (See comments.)
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise | (See comments.)
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Most of these are pretty minor things that will help going forward.
